backend/datasource/datasource.go

```diff
@@ -0,0 +1,104 @@
+package datasource
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
+)
+
+// Instance defines the interface for data source plugin instances.
+type Instance interface {
+	backend.CheckHealthHandler
+	backend.CallResourceHandler
+	backend.QueryDataHandler
+}
+
+// InstanceProviderFunc is the factory method for creating a new data source plugin instance.
+type InstanceProviderFunc func(settings backend.DataSourceInstanceSettings) (Instance, error)
+
+type Plugin interface {
+	ServeOpts() backend.ServeOpts
+	Serve() error
+}
+
+type plugin struct {
+	serveOpts backend.ServeOpts
+	im        instancemgmt.InstanceManager
+}
+
+func New(fn InstanceProviderFunc) Plugin {
+	if fn == nil {
+		panic("fn cannot be nil")
+	}
+
+	ip := NewInstanceProvider(func(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		instance, err := fn(settings)
+		if err != nil {
+			return nil, err
+		}
+
+		return instance, nil
+	})
+
+	p := &plugin{
+		im: instancemgmt.New(ip),
+	}
+	serveOpts := backend.ServeOpts{
+		CheckHealthHandler:  p,
+		CallResourceHandler: p,
+		QueryDataHandler:    p,
+	}
+	p.serveOpts = serveOpts
```

Comment:
Does putting the server config inside here make it harder to have a single instance support multiple "types" later?